Sources/Purchasing/StoreKit2/Win-Back Offers/WinBackOfferEligibilityCalculator.swift

```diff
@@ -0,0 +1,135 @@
+        return try await self.calculateEligibleWinBackOffers(forProduct: product)
+    }
+
+}
+
+// MARK: - Implementation
+@available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
+extension WinBackOfferEligibilityCalculator {
+
+    @available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
+    private func calculateEligibleWinBackOffers(
+        forProduct product: StoreProduct
+    ) async throws -> [WinBackOffer] {
+
+        #if compiler(>=6.0)
+        guard self.systemInfo.storeKitVersion.isStoreKit2EnabledAndAvailable else {
+            throw ErrorUtils.featureNotSupportedWithStoreKit1Error()
+        }
+
+        let eligibleWinBackOfferIDs: [String] = await self.calculateEligibleWinBackOfferIDs(forProduct: product)
+        guard !eligibleWinBackOfferIDs.isEmpty else { return [] }
+
+        guard let allWinBackOffersForThisProduct: [
+            Product.SubscriptionOffer
+        ] = product.sk2Product?.subscription?.winBackOffers else {
+            // StoreKit.Product.SubscriptionInfo is nil if the product is not a subscription
+            return []
+        }
+
+        let eligibleWinBackOffers = allWinBackOffersForThisProduct
+            // 1. Filter out the offers that the subscriber is not eligible for
+            .filter({
+                Set(eligibleWinBackOfferIDs).contains($0.id)
+            })
+            // 2. Convert the eligible offers to StoreProductDiscounts for us to use
+            .compactMap({
+                StoreProductDiscount(sk2Discount: $0, currencyCode: product.currencyCode)
+            })
+            // 3. Convert the StoreProductDiscounts to WinBackOffer objects
+            .map({
+                WinBackOffer(discount: $0)
+            })
+
+        return eligibleWinBackOffers
+        #else
+        return []
+        #endif
+    }
+
+    @available(iOS 18.0, macOS 15.0, tvOS 18.0, watchOS 11.0, visionOS 2.0, *)
+    private func calculateEligibleWinBackOfferIDs(forProduct product: StoreProduct) async -> [String] {
+        #if compiler(>=6.0)
+        guard let statuses = try? await product.sk2Product?.subscription?.status, !statuses.isEmpty else {
+            // If StoreKit.Product.subscription is nil, then the product isn't a subscription
+            // If statuses is empty, the subscriber was never subscribed to a product in the subscription group.
+            return []
+        }
+
+        let purchasedSubscriptionStatuses = statuses.filter({
+            switch $0.transaction {
+            case .unverified:
+                return false
+            case .verified(let transaction):
+                // Intentionally exclude transactions acquired through family sharing
+                return transaction.ownershipType == .purchased
+            }
+        })
+
+        let renewalInfos: [Product.SubscriptionInfo.RenewalInfo] = purchasedSubscriptionStatuses.compactMap({
```

Comment:
maybe we could add a little helper like we have for transactions:

https://github.com/RevenueCat/purchases-ios/blob/bcef02adff94f0aed2f72f4cb48f64d8a410b6cd/Sources/Purchasing/StoreKit2/StoreKit2TransactionFetcher.swift#L156

Reply:
Love this, added in 1a5630e3c827d18e2250939ba8fb2c4237e5f9ec